### DISCUSSION TITLE
How to handle new tab when we have dynamic url

### DISCUSSION THREAD

**Discussion by sandeepbaggi** [cypress-io/cypress#28955] [Questions and Help]
cy.get('myLink').click();

This link will open a new tab which has a dynamic url with no href and target attribute. All my automation needs to be done in the new tab which will be opened. 

The main page we only use for login and from there we will be redirected to new tab which is my localhost.

I have tried multiple solutions but nothing worked.

**Comment by portyu9** [comment]
Hi @sandeepbaggi 

The older `removeAttr('target')` workaround does not fit this case if the new tab is created dynamically and there is no `href`/`target` to modify.

For a **real multi-tab workflow**, the current supported Cypress approach is `@cypress/puppeteer`.

Install it with Puppeteer, then configure a message handler that finds the newly created page:

```ts
import { defineConfig } from 'cypress'
import { setup, retry } from '@cypress/puppeteer'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      setup({
        on,
        onMessage: {
          async switchToNewTab(browser) {
            const page = await retry(async () => {
              const pages = await browser.pages()

              const page = pages.find((p) =>
                p.url().startsWith('http://localhost')
              )

              if (!page) {
                throw new Error('New tab has not opened yet')
              }

              return page
            })

            await page.bringToFront()

            return page.url()
          },
        },
      })
    },
  },
})
```

Import its support commands once:

```js
import '@cypress/puppeteer/support'
```

Then:

```js
cy.get('myLink').click()

cy.puppeteer('switchToNewTab')
  .should('include', 'localhost')
```

The URL can be dynamic; you only need enough information to identify the new page.

One limitation remains: Cypress commands themselves do not switch into the second tab. Interactions with that tab are performed through Puppeteer's `page` API inside the Puppeteer handler.

Please mark this answer as accepted if it helped, thank you.
